llvm/lib/Target/AMDGPU/AMDGPUOccupancyAndLatencyHelper.cpp

```diff
@@ -0,0 +1,187 @@
+//==- AMDGPUOccupancyAndLatencyHelper.cpp - Helpers for occupancy + latency ==//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//==------------------------------------------------------------------------==//
+//
+/// \file
+/// \brief Helper functions for occupancy and latency.
+//
+//==------------------------------------------------------------------------==//
+
+#include "AMDGPUOccupancyAndLatencyHelper.h"
+#include "AMDGPUSubtarget.h"
+#include "GCNSubtarget.h"
+#include "SIInstrInfo.h"
+#include "SIRegisterInfo.h"
+
+#include "llvm/CodeGen/MachineLoopInfo.h"
+
+namespace llvm {
+
+// Other info which can help compare schedule result.
+float SchedScore::computeScore() const {
+  // Occupancy 1 cannot mix alu.
+  unsigned MixHidenAlu = Alu - MixAlu;
+  if (Occupancy == 1)
+    MixHidenAlu = 0;
+  return ((float)MemLatency - (float)MixHidenAlu) / (float)Occupancy -
+         LatencyHide;
+}
+float SchedScore::computeScore2() const {
+  float cycles = 0;
+  cycles = (MixAlu * Occupancy + MemLatency);
+  cycles /= Occupancy;
+  return cycles;
+}
+
+void SchedScore::sum(const SchedScore &s, unsigned loopDepth) {
+  unsigned loopCount = loopDepth > 0 ? std::pow(3, loopDepth) : 1;
```

Comment:
Need to `#include <cmath>` for this. But perhaps it would be better to use an integer-only power operation (not sure if we already have one anywhere in LLVM).